File: adaptor/go/redact.go
package microscope

import (
	"encoding/json"
	"strings"
)

var sensitiveKeys = map[string]struct{}{
	"password":         {},
	"password_hash":    {},
	"new_password":     {},
	"old_password":     {},
	"current_password": {},
	"refresh_token":    {},
	"access_token":     {},
	"token":            {},
	"otp":              {},
	"code":             {},
	"secret":           {},
	"encryption_key":   {},
	"authorization":    {},
	"mfa_secret":       {},
	"backup_codes":     {},
}

var sensitiveHeaderKeys = map[string]struct{}{
	"authorization": {},
	"cookie":        {},
	"x-api-key":     {},
	"x-auth-token":  {},
}

// RedactMap returns a copy of m with sensitive values masked.
func RedactMap(m map[string]any) map[string]any {
	if m == nil {
		return nil
	}
	out := make(map[string]any, len(m))
	for k, v := range m {
		out[k] = redactValue(k, v)
	}
	return out
}

// RedactHeaders returns a copy of headers with sensitive values masked.
func RedactHeaders(headers map[string][]string) map[string][]string {
	if headers == nil {
		return nil
	}
	out := make(map[string][]string, len(headers))
	for k, vals := range headers {
		if _, ok := sensitiveHeaderKeys[strings.ToLower(k)]; ok {
			out[k] = []string{"[REDACTED]"}
			continue
		}
		out[k] = append([]string(nil), vals...)
	}
	return out
}

func redactValue(key string, v any) any {
	if _, ok := sensitiveKeys[strings.ToLower(key)]; ok {
		return "[REDACTED]"
	}
	switch val := v.(type) {
	case map[string]any:
		return RedactMap(val)
	case []any:
		out := make([]any, len(val))
		for i, item := range val {
			out[i] = redactValue(key, item)
		}
		return out
	default:
		return v
	}
}

// RedactJSON parses JSON bytes, redacts sensitive fields, and re-marshals.
func RedactJSON(body []byte) string {
	if len(body) == 0 {
		return ""
	}
	var data any
	if err := json.Unmarshal(body, &data); err != nil {
		return string(body)
	}
	redacted := redactValue("", data)
	out, err := json.Marshal(redacted)
	if err != nil {
		return string(body)
	}
	return string(out)
}

File: adaptor/python/microscope/redact.py
from __future__ import annotations

import json
from typing import Any

SENSITIVE_KEYS = frozenset(
    {
        "password",
        "password_hash",
        "new_password",
        "old_password",
        "current_password",
        "refresh_token",
        "access_token",
        "token",
        "otp",
        "code",
        "secret",
        "encryption_key",
        "authorization",
        "mfa_secret",
        "backup_codes",
    }
)

SENSITIVE_HEADER_KEYS = frozenset(
    {
        "authorization",
        "cookie",
        "x-api-key",
        "x-auth-token",
    }
)


def redact_map(mapping: dict[str, Any] | None) -> dict[str, Any] | None:
    if mapping is None:
        return None
    return {key: _redact_value(key, value) for key, value in mapping.items()}


def redact_headers(headers: dict[str, list[str]] | None) -> dict[str, list[str]] | None:
    if headers is None:
        return None
    out: dict[str, list[str]] = {}
    for key, values in headers.items():
        if key.lower() in SENSITIVE_HEADER_KEYS:
            out[key] = ["[REDACTED]"]
        else:
            out[key] = list(values)
    return out


def redact_json(body: bytes) -> str:
    if not body:
        return ""
    try:
        data = json.loads(body)
    except json.JSONDecodeError:
        return body.decode("utf-8", errors="replace")
    redacted = _redact_value("", data)
    try:
        return json.dumps(redacted)
    except (TypeError, ValueError):
        return body.decode("utf-8", errors="replace")


def redact_args(args: list[Any]) -> list[Any] | None:
    if not args:
        return None
    out: list[Any] = []
    for arg in args:
        if isinstance(arg, dict):
            out.append(redact_map(arg))
        elif isinstance(arg, (bytes, bytearray)):
            out.append("[bytes]")
        else:
            out.append(arg)
    return out


def _redact_value(key: str, value: Any) -> Any:
    if isinstance(key, str) and key.lower() in SENSITIVE_KEYS:
        return "[REDACTED]"
    if isinstance(value, dict):
        return redact_map(value)
    if isinstance(value, list):
        return [_redact_value(key, item) for item in value]
    return value
